File: examples/benchmarks/2_full_scattering_speed/full_scattering.jl
using Pkg
Pkg.activate(dirname(@__DIR__))
Pkg.instantiate()

using BenchmarkTools
using CSV
using DataFrames
using ProgressBars
using ScatteringOptics
using VLBISkyModels

# create a scattering model
const sm = ScatteringModel()

# set the field of view
const dx = μas2rad(1)
const nx_list = [32, 64, 128, 256, 512, 1024, 2048]

# frequency
const νref = 230e9

# make a Gaussian Model
g = stretched(Gaussian(), dx*20, dx*20)

etime = []
atime = []
for nx in ProgressBar(nx_list)
    # get the model image
    fov = dx * nx
    grid = imagepixels(fov, fov, nx, nx)
    im = intensitymap(g, grid)

    # ensemble average
    out = @benchmark ensembleaverage($sm, $im; νref=νref)
    push!(etime, median(out).time/1e9)

    # full scattering
    out = @benchmark scatter_image($sm, $im; νref=νref)
    push!(atime, median(out).time/1e9)
end

df = DataFrame(
    nx = nx_list,
    ensembleaverage = etime,
    average = atime,
)
CSV.write("so_full_scattering.csv", df)

File: examples/benchmarks/2_full_scattering_speed/full_scattering.py
import ehtim as eh
import numpy as np
import pandas as pd
import timeit
import tqdm
import random

print("This script will benchmark eht-imaging's computation of scattering kernel.")

# numbe of samples
nx_list = [32, 64, 128, 256, 512, 1024, 2048]

# scattering models
sm = eh.scattering.ScatteringModel()

# dx
dx = 1 * eh.RADPERUAS

eatime = []
atime = []
for nx in tqdm.tqdm(nx_list):
    # set up an image
    fov = dx * nx
    im = eh.image.make_empty(nx, fov, 0.0, 0.0)
    im = im.add_gauss(1, [20 * dx, 20 * dx, 0.0, 0.0, 0.0])
    im.rf = 230e9

    def compute_ea():
        return sm.Ensemble_Average_Blur(im)

    def compute_a():
        return sm.Scatter(im)

    eatime.append(np.median(timeit.repeat(compute_ea, repeat=1, number=1)))
    atime.append(np.median(timeit.repeat(compute_a, repeat=1, number=1)))

df = pd.DataFrame()
df["nx"] = nx_list
df["ensembleaverage"] = eatime
df["average"] = atime
df.to_csv("eh_full_scattering.csv")
